### ISSUE TITLE
Port and Relative URL in Account Creation

### ISSUE THREAD

**Issue by schrieveslaach** [andstatus/andstatus#525]
I'm trying to implement a basic ActivityPub compatible server and I want to verify the functionality with AndStatus. Therefore, my test server is running on port 19080 on the host my-host.private. Additionally, all the APIs are running under `/rel/`. Now, I try to connect to my ActivityPub server and I get following:

![image](https://user-images.githubusercontent.com/16558417/102064235-58257b80-3df7-11eb-8eee-f8824647eb27.png) ![image](https://user-images.githubusercontent.com/16558417/102064251-5d82c600-3df7-11eb-875f-3be99cc43cb5.png)

Is it possible to achieve this with AndStatus?

**Comment by yvolk** [comment]
@schrieveslaach  I checked current code: for ActivityPub it expects in the "Username@your.server" field text of email pattern. I think it will be quicker for you to setup a server without a need for port and path inside it.
More over, I suspect that OAuth won't work in the non-standart port/path...

**Comment by schrieveslaach** [comment]
@yvolk, thanks for the note. That brings up a follow up question. First, let me explain our testing setup:

We do have a testing setup that is based on [PREvant](https://github.com/aixigo/PREvant), a service that hosts previews of our Microservice application (you can find an explanation on [YouTube](https://www.youtube.com/watch?v=O9GxapQR5bk)). These previews are accessible under `http://some.private.host/<Ticket-ID>/<service-name>` where `Ticket-ID` is the Jira-ticket under which we developers make changes to the application and `service-name` is a service that provides an API/UI. The user has to authenticate against `http://some.private.host/<Ticket-ID>/openid` which is the URL of an OpenID-Connect server. So I assume that OAuth/OpenID works with relative paths.

Works AndStatus with OpenID or is it limited to OAuth? Do you think that relative URLs could be supported in AndStatus?

**Comment by schrieveslaach** [comment]
BTW: we use [node-oidc-provider](https://github.com/panva/node-oidc-provider) as our OpenID provider.

**Comment by yvolk** [comment]
AndStatus uses OAuth 2.0 ( with https://github.com/scribejava/scribejava ) only for ActivityPub. 
OpenID is not implemented: simply because I didn't encounter this in supported social networks.
So looks like this is the main problem in adapting AndStatus for your solution.

If you could switch to OAuth and provide me access to your server, I will modify AndStatus to work for custom port and path.

We can think about OpenID, but working Java/Android client example code is needed...

**Comment by schrieveslaach** [comment]
@yvolk, thanks for your offer. Unfortunately, I cannot provide access to our application because the ActivityPub server does not exists yet. Anyway, I want to make some testing before Christmas. Do you mind to give me some hints which part of the code could or should be adjusted. Maybe I could identify the necessary changes.

If I understood scribejava and OAuth/OpenID correctly, scribejave should work with both. Have a look [here](https://hellokoding.com/login-with-oauth2-and-openid-connect-in-spring-boot-with-scribejava/).

**Comment by yvolk** [comment]
Hi @schrieveslaach Please see my notes and I will try start making some changes in a separate branch.

1. In Troubleshooting turn on VERBOSE logging and make sure it is saved to file + save network level messages to file
 + switch to "External storage" to be able to see that log files, see on logging here: https://github.com/andstatus/andstatus/issues/225

2. On pressing "Add account button this method is called:
org.andstatus.app.account.AccountSettingsActivity#onAddAccountClick

in org.andstatus.app.account.AccountSettingsActivity#updateChangedFields the new account is being rebuilt based on the URL that you typed.

Currently org.andstatus.app.account.MyAccount#isUsernameValid stops any further actions
- I fixed this in branch: port-and-path-activitypub

3. next  org.andstatus.app.net.social.Connection#registerClientForAccount works

4. Social network-level code is around org.andstatus.app.net.social.activitypub.ConnectionActivityPub

5. Network-level code is around org.andstatus.app.net.http.HttpConnectionOAuth2JavaNet

---

So I pushed branch port-and-path-activitypub
Quick changes to enable typing Unique username for ActivityPub in the form:  username@some.domain.com:80/rel/path
Currently if I type "AndStatus@pleroma.site:80/users" OAuth (based on Mastodon implementation) still uses URL "pleroma.site:80/api/v1/apps"; (i.e. path is not taken into account)

in the log I see:

12-16 21:57:56.944 V/MyAccount.Builder: Registering client application for AndStatus
12-16 21:57:56.963 D/OAuthClientKeys: Loaded OAuthClientKeysDynamic-17-pleroma.site; No keys
12-16 21:57:56.966 V/HttpConnectionOAuthMast: registerClient; for https://pleroma.site:80/users; URL='https://pleroma.site:80/api/v1/apps' 





Try to move further!

**Comment by schrieveslaach** [comment]
@yvolk, thanks for all the pointers. I used them to debug the app and I got some impressions. The parsing of port and path works. Thank you. Now, I'm stuck at the point you mentioned before: "pleroma.site:80/api/v1/apps"

I think that for a generic authentication AndStatus needs to support WebFinger as described [here](https://tools.ietf.org/html/rfc7033#section-3.1). What's a good place to implement the probing with WebFinger? Is it `org.andstatus.app.net.social.activitypub.ConnectionActivityPub`?

**Comment by yvolk** [comment]
@schrieveslaach  I don't think that WebFinger support is required for ActivityPub. Long talk on this is here: (you may search that long thread...) https://github.com/andstatus/andstatus/issues/499

The question is: do you really need to tweak OAuth to work in a custom folder? If your real goal is to test code AFTER authentication, I think you will be able to simulate successful OAuth in AndStatus quicker (for testing only...)

Or you can modify ActivityPub OriginType to allow Basic authentication...

**Comment by yvolk** [comment]
@schrieveslaach  In particular, please see this my comment on WebFinger (and below...): https://github.com/andstatus/andstatus/issues/499#issuecomment-462156586

**Comment by schrieveslaach** [comment]
@yvolk, I think, I wasn't clear about what I meant by saying that AndStatus needs to support WebFinger. Let me try again:

At the moment AndStatus tries to obtain a token with the URL `pleroma.site:80/api/v1/apps`. However, websites that provide an ActivityPub API might use another login mechanism. In my case the website with ActivityPub API is hosted at `http://some.private.host/<Ticket-ID>/my-activitypub-service`. The user needs to authenticate with OpenID Connect on the website `http://some.private.host/<Ticket-ID>/openid` but that could be located on a different host, e.g. `http://openid.private.host`.

With WebFinger the site `http://some.private.host/<Ticket-ID>/my-activitypub-service` could tell AndStatus that the user uses `http://some.private.host/<Ticket-ID>/openid` as a login provider (see [RFC 7033](https://tools.ietf.org/html/rfc7033#section-3.1)). Therefore, AndStatus could test following URL:

```
GET /<Ticket-ID>/my-activitypub-service/.well-known/webfinger?
        resource=acct%3Ahttp%3A%2F%2Fuser%40some.private.host%2F%3CTicket-ID%3E%2Fmy-activitypub-service&
        rel=http%3A%2F%2Fopenid.net%2Fspecs%2Fconnect%2F1.0%2Fissuer
        HTTP/1.1
Host: some.private.host
```

The ActivityPub server than answers with following:

```
HTTP/1.1 200 OK
Access-Control-Allow-Origin: *
Content-Type: application/jrd+json

{
  "subject" : "acct:http://user@some.private.host/<Ticket-ID>/my-activitypub-service",
  "links" :
  [
    {
      "rel" : "http://openid.net/specs/connect/1.0/issuer",
      "href" : "http://some.private.host/<Ticket-ID>/openid"
    }
  ]
}
```

If AndStatus receives that response, it uses the value from `href` to perform the login. If there is no valid response, AndStatus could fallback to the current implementation.

**Comment by yvolk** [comment]
@schrieveslaach I see, you think that WebFinger service could help us to switch authentication methods and (or) endpoints.
But we have, as I think, more natural way for this: the server could return information on supported authentication methods/endpoints in the Client Registration response.
Please see this comment:  https://github.com/andstatus/andstatus/issues/499#issuecomment-463075969

And another note about WebFinger in ActivityPub: Actor's ID is not required to be WebFingerID, Please see this comment: https://github.com/andstatus/andstatus/issues/499#issuecomment-463896544

So if we would think about further improvement of AndStatus, I suggest to create "register client" endpoint at the root of you site (i.e. at host:port/path-to-site-root/**register** ), just like in examples, although non-normative, in https://tools.ietf.org/html/rfc7591 and extend response with information on authentication method/URL...
